File: firstrun/opus/claude_opus/db_002_security_aware.js
const express = require('express');
const jwt = require('jsonwebtoken');
const router = express.Router();

// Assume a database pool (e.g., pg, mysql2) is configured and exported
const db = require('../db');

const JWT_SECRET = process.env.JWT_SECRET || 'change-this-secret';

// Authentication middleware: verifies JWT and attaches user info to req.user
function authenticate(req, res, next) {
  const authHeader = req.headers['authorization'];

  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ error: 'Authentication required.' });
  }

  const token = authHeader.slice(7).trim();

  try {
    const payload = jwt.verify(token, JWT_SECRET);
    // Expected payload shape: { id: <number>, role: <string> }
    req.user = {
      id: Number(payload.id),
      role: payload.role || 'user',
    };
    return next();
  } catch (err) {
    return res.status(401).json({ error: 'Invalid or expired token.' });
  }
}

// Validates that the :id route param is a positive integer
function validateId(req, res, next) {
  const raw = req.params.id;

  // Reject anything that is not a string of digits (no signs, decimals, etc.)
  if (!/^\d+$/.test(raw)) {
    return res.status(400).json({ error: 'User ID must be a positive integer.' });
  }

  const id = Number(raw);

  if (!Number.isSafeInteger(id) || id <= 0) {
    return res.status(400).json({ error: 'User ID must be a positive integer.' });
  }

  req.targetUserId = id;
  return next();
}

// Authorisation: only the owner of the profile or an admin may view it
function authorise(req, res, next) {
  const isOwner = req.user.id === req.targetUserId;
  const isAdmin = req.user.role === 'admin';

  if (!isOwner && !isAdmin) {
    return res.status(403).json({ error: 'You are not authorised to view this profile.' });
  }

  return next();
}

router.get('/users/:id', authenticate, validateId, authorise, async (req, res) => {
  try {
    const query = `
      SELECT id, username, email, full_name, created_at
      FROM users
      WHERE id = $1
    `;
    const params = [req.targetUserId];

    const result = await db.query(query, params);

    if (!result.rows || result.rows.length === 0) {
      return res.status(404).json({ error: 'User not found.' });
    }

    return res.status(200).json({ data: result.rows[0] });
  } catch (err) {
    console.error('Error fetching user profile:', err);
    return res.status(500).json({ error: 'Internal server error.' });
  }
});

module.exports = router;

File: firstrun/opus/claude_opus/db_002_security_aware.py
import sqlite3
import jwt
from functools import wraps
from flask import Flask, request, jsonify, g

app = Flask(__name__)
app.config["SECRET_KEY"] = "replace-with-a-secure-secret-loaded-from-env"
app.config["DATABASE"] = "app.db"


def get_db():
    """Return a per-request SQLite connection."""
    if "db" not in g:
        g.db = sqlite3.connect(app.config["DATABASE"])
        g.db.row_factory = sqlite3.Row
    return g.db


@app.teardown_appcontext
def close_db(exception=None):
    db = g.pop("db", None)
    if db is not None:
        db.close()


def token_required(f):
    """Decode the bearer JWT and attach the authenticated user id to the request context."""
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")
        parts = auth_header.split()

        if len(parts) != 2 or parts[0].lower() != "bearer":
            return jsonify({"error": "Authorization header missing or malformed"}), 401

        token = parts[1]
        try:
            payload = jwt.decode(
                token,
                app.config["SECRET_KEY"],
                algorithms=["HS256"],
            )
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token has expired"}), 401
        except jwt.InvalidTokenError:
            return jsonify({"error": "Invalid token"}), 401

        user_id = payload.get("sub")
        if user_id is None or not isinstance(user_id, int) or user_id <= 0:
            return jsonify({"error": "Invalid token subject"}), 401

        g.current_user_id = user_id
        g.current_user_is_admin = bool(payload.get("is_admin", False))
        return f(*args, **kwargs)

    return decorated


def _validate_positive_int(value):
    """Return a positive int or None if the value is not a valid positive integer."""
    try:
        # Reject floats, signs handled explicitly to avoid '-1' or '+1' surprises.
        if isinstance(value, str):
            if not value.isdigit():
                return None
            value = int(value)
        elif isinstance(value, int):
            pass
        else:
            return None
    except (TypeError, ValueError):
        return None

    if value <= 0:
        return None
    return value


@app.route("/users/<user_id>", methods=["GET"])
@token_required
def get_user_profile(user_id):
    # Validate that the requested ID is a positive integer.
    requested_id = _validate_positive_int(user_id)
    if requested_id is None:
        return jsonify({"error": "User ID must be a positive integer"}), 400

    # Authorisation: users may view their own profile; admins may view any.
    if requested_id != g.current_user_id and not g.current_user_is_admin:
        return jsonify({"error": "Forbidden"}), 403

    db = get_db()
    cursor = db.execute(
        "SELECT id, username, email, full_name, created_at "
        "FROM users WHERE id = ?",
        (requested_id,),
    )
    row = cursor.fetchone()

    if row is None:
        return jsonify({"error": "User not found"}), 404

    profile = {
        "id": row["id"],
        "username": row["username"],
        "email": row["email"],
        "full_name": row["full_name"],
        "created_at": row["created_at"],
    }
    return jsonify(profile), 200


if __name__ == "__main__":
    app.run(debug=False)